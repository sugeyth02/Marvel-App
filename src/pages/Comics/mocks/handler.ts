import { rest } from 'msw';

const baseURL = 'https://gateway.marvel.com/v1/public/';

export const handlers = [
  rest.get(`${baseURL}comics`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        data: {
          count: 1560,
          results: [
            {
              id: 82967,
              title: 'Marvel Previews (2017)',
              thumbnail: {
                path: 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available',
              },
            },
            {
              id: 376,
              title: 'Ant-Man (2003) #3',
              thumbnail: {
                path: 'http://i.annihil.us/u/prod/marvel/i/mg/d/70/4bc69c7e9b9d7',
              },
            },
          ],
        },
      })
    );
  }),
];
