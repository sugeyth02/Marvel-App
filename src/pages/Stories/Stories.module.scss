@use './../../variables' as variables;

.container {
  padding: 60px 20px 20px 20px;
}
.container__header {
  font-family: variables.$font-stack;
  font-weight: 500;
  font-size: 1.5rem;
  padding-bottom: 20px;
}
.container__filters {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-end;
  padding: 30px 0;
  width: 100%;

  label {
    font-family: variables.$font-stack;
    padding: 10px;
    font-size: 1rem;
    font-weight: 500;
  }
  select {
    padding: 4px;
    font-family: variables.$font-stack;
    font-size: 0.9rem;
  }
}

@media (max-width: 445px) {
  .container__filters {
    justify-content: flex-start;
  }
  .container__filters {
    select {
      width: 40%;
    }
  }
}
