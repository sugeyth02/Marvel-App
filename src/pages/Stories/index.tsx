import React, { useCallback, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import Item from '../../components/Item';
import Pagination from '../../components/Pagination';
import { ICharacter } from '../../entities/ICharacter';
import { IStorie } from '../../entities/IStorie';
import services from '../../services/service';
import styles from './Stories.module.scss';

export default function Stories() {
  const [page, setPage] = useState(1);
  const [countItems, setCountItems] = useState(0);
  const [stories, setStories] = useState<IStorie[]>([]);
  const [characters, setCharacters] = useState<ICharacter[]>([]);
  const [search, setSearch] = useSearchParams();
  const [pageQuery, characterQuery] = [
    search.get('page'),
    search.get('character'),
  ];

  useEffect(() => {
    const getData = async () => {
      try {
        const response = await services.getStories({
          limit: 30,
          ...(pageQuery !== null && {
            offset: `${(Number(pageQuery) - 1) * 30}`,
          }),
          ...(characterQuery !== null && { characters: `${characterQuery}` }),
        });
        setStories(response.result);
        setCountItems(response.count);
      } catch (err) {}
    };
    getData();
  }, [characterQuery, pageQuery]);

  useEffect(() => {
    const getData = async () => {
      try {
        const response = await services.getCharacters({ limit: 100 });
        setCharacters(response.result);
      } catch (err) {}
    };
    getData();
  }, []);

  const handleCharacterSelected = useCallback(
    (value: string) => {
      setSearch({
        page: '1',
        character: `${value}`,
      });
    },
    [setSearch]
  );

  useEffect(() => {
    setPage(1);
  }, [characterQuery, setPage]);
  
  useEffect(() => {
    setSearch({
      page: `${page}`,
      ...(characterQuery !== null && { character: characterQuery }),
    });
  }, [page, characterQuery, setSearch]);

  return (
    <div className={styles.container}>
      <h1 className={styles.container__header}>MARVEL STORIES LIST</h1>
      <div className={styles.container__filters}>
        <div className={styles.filters__select}>
          <label htmlFor='characters'>Characters:</label>
          <select
            name='characters'
            id='characters'
            onChange={(e) => {
              handleCharacterSelected(e.target.value);
            }}
          >
            {characters.map((character) => (
              <option value={character.id} key={character.id}>
                {character.name}
              </option>
            ))}
          </select>
        </div>
      </div>
      <div className={styles.container__items}>
        {stories.map((story) => (
          <Item
            page='stories'
            title={story.title}
            id={story.id}
            key={story.id}
          />
        ))}
      </div>
      <Pagination
        setPage={setPage}
        currentPage={page}
        totalItems={countItems}
        itemsperPage={30}
      />
    </div>
  );
}
