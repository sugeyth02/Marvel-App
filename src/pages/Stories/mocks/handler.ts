import { rest } from 'msw';

const baseURL = 'https://gateway.marvel.com/v1/public/';

export const handlers = [
  rest.get(`${baseURL}stories`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        data: {
          count: 1560,
          results: [
            {
              id: 7,
              title: 'Investigating the murder of a teenage girl, Cage suddenly learns that a three-way gang war is under way for control of the turf',
            },
            {
              id: 40,
              title: 'Ever-expanding their ranks, the Children of the Atom combat the evils threatening both mutants and humans Ã‘ like the Brotherhood',
            },
          ],
        },
      })
    );
  }),

  rest.get(`${baseURL}characters`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        data: {
          results: [
            {
              id: 1011334,
              name: '3-D Man',
            },
            {
              id: 1009174,
              name: 'Beak',
            },
          ],
        },
      })
    );
  }),
];
