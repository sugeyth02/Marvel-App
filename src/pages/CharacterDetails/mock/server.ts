import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { handlers as characterHandler } from '../../Characters/mock/handler';

const server = setupServer(...handlers,...characterHandler);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
