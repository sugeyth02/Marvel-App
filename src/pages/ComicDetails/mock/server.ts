import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { handlers as comicsHandler } from '../../Comics/mocks/handler';

const server = setupServer(...handlers, ...comicsHandler);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
