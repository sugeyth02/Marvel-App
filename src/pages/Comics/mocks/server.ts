import { setupServer } from 'msw/node';
import { handlers } from './handler';


const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
