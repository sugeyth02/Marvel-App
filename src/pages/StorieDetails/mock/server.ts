import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { handlers as storiesHandler } from '../../Stories/mocks/handler';

const server = setupServer(...handlers, ...storiesHandler);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
