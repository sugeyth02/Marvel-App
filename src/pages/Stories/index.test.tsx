import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { store } from './../../redux/store';
import { Provider } from 'react-redux';
import userEvent from '@testing-library/user-event';
import './mocks/server';
import Stories from '.';

describe('Testing Stories list view', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  test('should renders a list of stories initially', async () => {
    // Render
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Stories />
        </BrowserRouter>
      </Provider>
    );

    // Assertions
    await waitFor(
      () => {
        expect(
          screen.queryByText(
            'Investigating the murder of a teenage girl, Cage suddenly learns that a three-way gang war is under way for control of the turf'
          )
        ).toBeInTheDocument();
        expect(
          screen.queryByText(
            'Ever-expanding their ranks, the Children of the Atom combat the evils threatening both mutants and humans Ã‘ like the Brotherhood'
          )
        ).toBeInTheDocument();
        expect(container).toMatchSnapshot();
      },
      { timeout: 3000 }
    );

    const characters = await screen.findByLabelText(/Characters:/i);
    userEvent.click(characters);

    await waitFor(() => {
      expect(screen.queryByText('3-D Man')).toBeInTheDocument();
      expect(screen.queryByText('Beak')).toBeInTheDocument();
    });
  });
});
