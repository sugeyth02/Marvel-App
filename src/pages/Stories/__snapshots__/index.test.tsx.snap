// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Stories list view should renders a list of stories initially 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="container__header"
    >
      MARVEL STORIES LIST
    </h1>
    <div
      class="container__filters"
    >
      <div
        class="filters__select"
      >
        <label
          for="characters"
        >
          Characters:
        </label>
        <select
          id="characters"
          name="characters"
        />
      </div>
    </div>
    <div
      class="container__items"
    >
      <li
        class="item"
      >
        Investigating the murder of a teenage girl, Cage suddenly learns that a three-way gang war is under way for control of the turf
      </li>
      <li
        class="item"
      >
        Ever-expanding their ranks, the Children of the Atom combat the evils threatening both mutants and humans Ã‘ like the Brotherhood
      </li>
    </div>
    <div
      class="container"
    >
      <button
        class="container__button"
        disabled=""
      >
        <i
          class="fa-solid fa-angle-left"
        />
      </button>
      <div
        class="numbers"
      />
      <button
        class="container__button"
      >
        <i
          class="fa-solid fa-angle-right"
        />
      </button>
    </div>
  </div>
</div>
`;
